File: game_catalog.go
// game_catalog.go
package main

import (
	"encoding/csv"
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

const dataFile = "games.json"

type Game struct {
	Name   string  `json:"name"`
	Genre  string  `json:"genre"`
	Hours  float64 `json:"hours"`
	Price  float64 `json:"price"`
	Rating float64 `json:"rating"`
}

type Catalog struct {
	Games []Game `json:"games"`
}

func (c *Catalog) load() {
	data, err := os.ReadFile(dataFile)
	if err != nil {
		c.Games = []Game{}
		return
	}
	if err := json.Unmarshal(data, c); err != nil {
		c.Games = []Game{}
	}
}

func (c *Catalog) save() {
	data, _ := json.MarshalIndent(c, "", "  ")
	os.WriteFile(dataFile, data, 0644)
}

func (c *Catalog) addGame(name, genre string, hours, price, rating float64) bool {
	for _, g := range c.Games {
		if strings.EqualFold(g.Name, name) {
			fmt.Printf("\033[31mИгра '%s' уже существует.\033[0m\n", name)
			return false
		}
	}
	c.Games = append(c.Games, Game{name, genre, hours, price, rating})
	c.save()
	fmt.Printf("\033[32mИгра '%s' добавлена.\033[0m\n", name)
	return true
}

func (c *Catalog) removeGame(name string) bool {
	for i, g := range c.Games {
		if strings.EqualFold(g.Name, name) {
			c.Games = append(c.Games[:i], c.Games[i+1:]...)
			c.save()
			fmt.Printf("\033[33mИгра '%s' удалена.\033[0m\n", name)
			return true
		}
	}
	fmt.Printf("\033[31mИгра '%s' не найдена.\033[0m\n", name)
	return false
}

func (c *Catalog) listGames(filterGenre, sortBy string) {
	games := c.Games
	if filterGenre != "" {
		filtered := []Game{}
		for _, g := range games {
			if strings.EqualFold(g.Genre, filterGenre) {
				filtered = append(filtered, g)
			}
		}
		if len(filtered) == 0 {
			fmt.Printf("\033[33mИгры с жанром '%s' не найдены.\033[0m\n", filterGenre)
			return
		}
		games = filtered
	}
	switch sortBy {
	case "hours":
		sort.Slice(games, func(i, j int) bool { return games[i].Hours > games[j].Hours })
	case "price":
		sort.Slice(games, func(i, j int) bool { return games[i].Price > games[j].Price })
	case "rating":
		sort.Slice(games, func(i, j int) bool { return games[i].Rating > games[j].Rating })
	default:
		sort.Slice(games, func(i, j int) bool { return games[i].Name < games[j].Name })
	}
	if len(games) == 0 {
		fmt.Println("\033[33mКаталог пуст.\033[0m")
		return
	}
	fmt.Println("\033[36m📋 Каталог игр:\033[0m")
	for _, g := range games {
		fmt.Printf("  \033[32m%s\033[0m | \033[33m%s\033[0m | \033[34m%.1f ч\033[0m | \033[35m%.2f руб\033[0m | \033[31m★ %.1f\033[0m\n",
			g.Name, g.Genre, g.Hours, g.Price, g.Rating)
	}
}

func (c *Catalog) searchGames(query string) {
	results := []Game{}
	for _, g := range c.Games {
		if strings.Contains(strings.ToLower(g.Name), strings.ToLower(query)) ||
			strings.Contains(strings.ToLower(g.Genre), strings.ToLower(query)) {
			results = append(results, g)
		}
	}
	if len(results) == 0 {
		fmt.Printf("\033[33mПо запросу '%s' ничего не найдено.\033[0m\n", query)
		return
	}
	fmt.Printf("\033[36m🔍 Найдено %d игр:\033[0m\n", len(results))
	for _, g := range results {
		fmt.Printf("  \033[32m%s\033[0m | \033[33m%s\033[0m | \033[34m%.1f ч\033[0m | \033[35m%.2f руб\033[0m | \033[31m★ %.1f\033[0m\n",
			g.Name, g.Genre, g.Hours, g.Price, g.Rating)
	}
}

func (c *Catalog) exportJSON(filename string) {
	data, _ := json.MarshalIndent(c, "", "  ")
	os.WriteFile(filename, data, 0644)
	fmt.Printf("\033[32mЭкспортировано в %s (JSON)\033[0m\n", filename)
}

func (c *Catalog) exportCSV(filename string) {
	f, _ := os.Create(filename)
	defer f.Close()
	w := csv.NewWriter(f)
	defer w.Flush()
	w.Write([]string{"name", "genre", "hours", "price", "rating"})
	for _, g := range c.Games {
		w.Write([]string{g.Name, g.Genre, strconv.FormatFloat(g.Hours, 'f', 1, 64),
			strconv.FormatFloat(g.Price, 'f', 2, 64), strconv.FormatFloat(g.Rating, 'f', 1, 64)})
	}
	fmt.Printf("\033[32mЭкспортировано в %s (CSV)\033[0m\n", filename)
}

func main() {
	var (
		add     string
		genre   string
		hours   float64
		price   float64
		rating  float64
		remove  string
		list    bool
		filter  string
		sortBy  string
		search  string
		expJson string
		expCsv  string
	)
	flag.StringVar(&add, "add", "", "Добавить игру")
	flag.StringVar(&genre, "genre", "", "Жанр")
	flag.Float64Var(&hours, "hours", 0, "Часы игры")
	flag.Float64Var(&price, "price", 0, "Цена")
	flag.Float64Var(&rating, "rating", 0, "Рейтинг (1-10)")
	flag.StringVar(&remove, "remove", "", "Удалить игру")
	flag.BoolVar(&list, "list", false, "Показать все игры")
	flag.StringVar(&filter, "filter", "", "Фильтр по жанру")
	flag.StringVar(&sortBy, "sort", "name", "Сортировка: name, hours, price, rating")
	flag.StringVar(&search, "search", "", "Поиск")
	flag.StringVar(&expJson, "export-json", "", "Экспорт в JSON")
	flag.StringVar(&expCsv, "export-csv", "", "Экспорт в CSV")
	flag.Parse()

	catalog := &Catalog{}
	catalog.load()

	if add != "" {
		if genre == "" || hours == 0 || price == 0 || rating == 0 {
			fmt.Println("\033[31mДля добавления игры требуются --genre, --hours, --price, --rating\033[0m")
			os.Exit(1)
		}
		if rating < 1 || rating > 10 {
			fmt.Println("\033[31mРейтинг должен быть от 1 до 10\033[0m")
			os.Exit(1)
		}
		catalog.addGame(add, genre, hours, price, rating)
	} else if remove != "" {
		catalog.removeGame(remove)
	} else if list {
		catalog.listGames(filter, sortBy)
	} else if search != "" {
		catalog.searchGames(search)
	} else if expJson != "" {
		catalog.exportJSON(expJson)
	} else if expCsv != "" {
		catalog.exportCSV(expCsv)
	} else {
		fmt.Println("Используйте --help для справки.")
	}
}

File: game_catalog.py
#!/usr/bin/env python3
# game_catalog.py
import argparse
import json
import csv
import sys
import os
from datetime import datetime
from colorama import init, Fore, Style

init(autoreset=True)

DATA_FILE = "games.json"

class Game:
    def __init__(self, name, genre, hours, price, rating):
        self.name = name
        self.genre = genre
        self.hours = hours
        self.price = price
        self.rating = rating

    def to_dict(self):
        return {"name": self.name, "genre": self.genre, "hours": self.hours,
                "price": self.price, "rating": self.rating}

    @classmethod
    def from_dict(cls, data):
        return cls(data["name"], data["genre"], data["hours"], data["price"], data["rating"])

class GameCatalog:
    def __init__(self):
        self.games = []
        self.load()

    def load(self):
        if not os.path.exists(DATA_FILE):
            self.games = []
            return
        try:
            with open(DATA_FILE, 'r', encoding='utf-8') as f:
                data = json.load(f)
                self.games = [Game.from_dict(item) for item in data]
        except (json.JSONDecodeError, FileNotFoundError):
            self.games = []

    def save(self):
        with open(DATA_FILE, 'w', encoding='utf-8') as f:
            json.dump([g.to_dict() for g in self.games], f, ensure_ascii=False, indent=2)

    def add_game(self, name, genre, hours, price, rating):
        if any(g.name.lower() == name.lower() for g in self.games):
            print(Fore.RED + f"Игра '{name}' уже существует.")
            return False
        game = Game(name, genre, hours, price, rating)
        self.games.append(game)
        self.save()
        print(Fore.GREEN + f"Игра '{name}' добавлена.")
        return True

    def remove_game(self, name):
        for i, g in enumerate(self.games):
            if g.name.lower() == name.lower():
                del self.games[i]
                self.save()
                print(Fore.YELLOW + f"Игра '{name}' удалена.")
                return True
        print(Fore.RED + f"Игра '{name}' не найдена.")
        return False

    def list_games(self, filter_genre=None, sort_by="name"):
        games = self.games
        if filter_genre:
            games = [g for g in games if g.genre.lower() == filter_genre.lower()]
            if not games:
                print(Fore.YELLOW + f"Игры с жанром '{filter_genre}' не найдены.")
                return
        if sort_by == "name":
            games.sort(key=lambda g: g.name.lower())
        elif sort_by == "hours":
            games.sort(key=lambda g: g.hours, reverse=True)
        elif sort_by == "price":
            games.sort(key=lambda g: g.price, reverse=True)
        elif sort_by == "rating":
            games.sort(key=lambda g: g.rating, reverse=True)
        else:
            games.sort(key=lambda g: g.name.lower())

        if not games:
            print(Fore.YELLOW + "Каталог пуст.")
            return
        print(Fore.CYAN + "📋 Каталог игр:")
        for g in games:
            print(f"  {Fore.GREEN}{g.name}{Style.RESET_ALL} | {Fore.YELLOW}{g.genre}{Style.RESET_ALL} | "
                  f"{Fore.BLUE}{g.hours} ч{Style.RESET_ALL} | {Fore.MAGENTA}{g.price:.2f} руб{Style.RESET_ALL} | "
                  f"{Fore.RED}★ {g.rating}{Style.RESET_ALL}")

    def search_games(self, query):
        results = [g for g in self.games if query.lower() in g.name.lower() or query.lower() in g.genre.lower()]
        if not results:
            print(Fore.YELLOW + f"По запросу '{query}' ничего не найдено.")
            return
        print(Fore.CYAN + f"🔍 Найдено {len(results)} игр:")
        for g in results:
            print(f"  {Fore.GREEN}{g.name}{Style.RESET_ALL} | {Fore.YELLOW}{g.genre}{Style.RESET_ALL} | "
                  f"{Fore.BLUE}{g.hours} ч{Style.RESET_ALL} | {Fore.MAGENTA}{g.price:.2f} руб{Style.RESET_ALL} | "
                  f"{Fore.RED}★ {g.rating}{Style.RESET_ALL}")

    def export_json(self, filename):
        with open(filename, 'w', encoding='utf-8') as f:
            json.dump([g.to_dict() for g in self.games], f, ensure_ascii=False, indent=2)
        print(Fore.GREEN + f"Экспортировано в {filename} (JSON)")

    def export_csv(self, filename):
        with open(filename, 'w', newline='', encoding='utf-8') as f:
            writer = csv.DictWriter(f, fieldnames=["name", "genre", "hours", "price", "rating"])
            writer.writeheader()
            for g in self.games:
                writer.writerow(g.to_dict())
        print(Fore.GREEN + f"Экспортировано в {filename} (CSV)")

def main():
    parser = argparse.ArgumentParser(description="Каталог игр (Steam)")
    parser.add_argument("--add", help="Добавить игру")
    parser.add_argument("--genre", help="Жанр")
    parser.add_argument("--hours", type=float, help="Часы игры")
    parser.add_argument("--price", type=float, help="Цена")
    parser.add_argument("--rating", type=float, help="Рейтинг (1-10)")
    parser.add_argument("--remove", help="Удалить игру")
    parser.add_argument("--list", action="store_true", help="Показать все игры")
    parser.add_argument("--filter", help="Фильтр по жанру")
    parser.add_argument("--sort", choices=["name", "hours", "price", "rating"], default="name", help="Сортировка")
    parser.add_argument("--search", help="Поиск по названию или жанру")
    parser.add_argument("--export-json", help="Экспорт в JSON")
    parser.add_argument("--export-csv", help="Экспорт в CSV")
    args = parser.parse_args()

    catalog = GameCatalog()

    if args.add:
        if not args.genre or args.hours is None or args.price is None or args.rating is None:
            print(Fore.RED + "Для добавления игры требуются --genre, --hours, --price, --rating")
            sys.exit(1)
        if args.rating < 1 or args.rating > 10:
            print(Fore.RED + "Рейтинг должен быть от 1 до 10")
            sys.exit(1)
        catalog.add_game(args.add, args.genre, args.hours, args.price, args.rating)

    elif args.remove:
        catalog.remove_game(args.remove)

    elif args.list:
        catalog.list_games(args.filter, args.sort)

    elif args.search:
        catalog.search_games(args.search)

    elif args.export_json:
        catalog.export_json(args.export_json)

    elif args.export_csv:
        catalog.export_csv(args.export_csv)

    else:
        parser.print_help()

if __name__ == "__main__":
    main()
